### PR TITLE
Add new algorithm: BISQ .wallet (scrypt) (-m 29800)

### PR DESCRIPTION
**Pull request by Banaanhangwagen**
Please find here a pull request in order to implement cracking a Bisq-wallet (scrypt). See [this issue](https://github.com/hashcat/hashcat/issues/3462) for more details.

It is mainly a copy from -m 27700, so no big changes...
Tested on my Linux-environment, and it works like a charm. ;)

Feel free to change little errors.

Thank you

### CODE REVIEW

File: tools/bisq2hashcat.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        iv = binascii.hexlify(iv).decode("ascii")
+        block_iv = binascii.hexlify(block_iv).decode("ascii")
+        block_noiv = binascii.hexlify(block_noiv).decode("ascii")
+        sys.stdout.write("%s:$bisq$%d*%s*%s*%s\n" % (bname, version, iv, block_iv, block_noiv))
+        return
+
+    f.close()
+
+if __name__ == "__main__":
+    if len(sys.argv) < 2:
```
Comment (ventaquil) [inline tools/bisq2hashcat.py:121]: Cannot we use [`argparse.ArgumentParser`](https://docs.python.org/3/library/argparse.html) here?

File: tools/bisq2hashcat.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        sys.stderr.write("Usage: %s [Bisq wallet file (.wallet)]\n" % sys.argv[0])
+ #       sys.stderr.write("\nMultiBit Classic -> for a wallet named 'xyz', we need the xyz-data/key-backup/xyz*.key OR xyz-data/wallet-backup/xyz*.wallet file\n")
+        sys.exit(1)
+
+    for j in range(1, len(sys.argv)):
+        process_file(sys.argv[j])
```
Comment (ventaquil) [inline tools/bisq2hashcat.py:127]:
```suggestion
    for arg in sys.argv[1:]:
        process_file(arg)
```

File: tools/bisq2hashcat.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    except IOError:
+        e = sys.exc_info()[1]
+        sys.stderr.write("%s\n" % str(e))
+        return
+
+    if "wallet" in bname or b"org.bitcoin.production" in data:
+ #       sys.stderr.write("[WARNING] Cracking .wallet files is a very slow process, try cracking the associated .key file instead!\n")
+        version = 3  # MultiBit Classic .wallet file
+        # def is_wallet_file(wallet_file) from btcrecover
+        wallet_file = open(filename, "rb")
```
Comment (ventaquil) [inline tools/bisq2hashcat.py:46]: Are we opening this file twice (`f` handle in 35th line is not closed)?

File: tools/bisq2hashcat.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        wallet_file.seek(0)
```
Comment (ventaquil) [inline tools/bisq2hashcat.py:47]: Do we need to do `seek(0)` when we open file?

File: tools/bisq2hashcat.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        is_valid_bitcoinj_wallet = False
+        if wallet_file.read(1) == b"\x0a":  # protobuf field number 1 of type length-delimited
```
Comment (ventaquil) [inline tools/bisq2hashcat.py:49]: I prefer to use [`with` statement](https://docs.python.org/3/reference/compound_stmts.html#with).

```suggestion
        with open(filename, "rb") as wallet_file:
            wallet_file.seek(0)
            is_valid_bitcoinj_wallet = False
            if wallet_file.read(1) == b"\x0a":  # protobuf field number 1 of type length-delimited
```

See also https://stackoverflow.com/a/31334226.

File: tools/bisq2hashcat.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+            network_identifier_len = ord(wallet_file.read(1))
+            if 1 <= network_identifier_len < 128:
+                wallet_file.seek(2 + network_identifier_len)
+                c = wallet_file.read(1)
+                if c and c in b"\x12\x1a":   # field number 2 or 3 of type length-delimited
+                    is_valid_bitcoinj_wallet = True
+        if is_valid_bitcoinj_wallet:
+            pb_wallet = Wallet()
+            pb_wallet.ParseFromString(data)
```
Comment (ventaquil) [inline tools/bisq2hashcat.py:58]: It's better to check for `not is_valid_bitcoinj_wallet` and then return early - one indent level less at all.

```suggestion
        if not is_valid_bitcoinj_wallet:
            return

        pb_wallet = Wallet()
        pb_wallet.ParseFromString(data)
```

File: tools/bisq2hashcat.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+from protobuf.wallet_pb2 import *
+
+
+def process_file(filename):
+    bname = os.path.basename(filename)
+    try:
+        f = open(filename, "rb")
+        data = f.read()
+    except IOError:
+        e = sys.exc_info()[1]
```
Comment (ventaquil) [inline tools/bisq2hashcat.py:38]:
```suggestion
    except IOError as e:
```

**Review by ventaquil — COMMENTED** [review]
I am not sure which Python version is target version for `tools/bisq2hashcat.py` (it looks like to support both 2.7 and 3.x version).

However we should keep in mind that Python 2.7 is not supported anymore - [Sunsetting Python 2](https://www.python.org/doc/sunset-python-2/).

I left few comments but I could have many more. If it's working I can add improvements in different PR to make this one clean.

**Comment by jsteube** [comment]
I'm merging this as-as, @ventaquil feel free to add another PR for the extraction tool changes.

**Pull request merged**